### DISCUSSION TITLE
Method to forget nested keys in collections

### DISCUSSION THREAD

**Discussion by aschmelyun** [laravel/framework#40749] [Ideas]
Hi all! Just wanted to see if this would be a desired addition. For collections, there's the `forget()` method that removes a specific key from a collection:

```php
$collection = collect(['name' => 'taylor', 'framework' => 'laravel']);

$collection->forget('name');

// ['framework' => 'laravel']
```

However, if there's a collection with nested elements, forget needs some kind of traversal alongside of it. Instead, I was proposing adding a new collection method to more easily handle this:

```php
$collection = collect([
    [
        'name' => 'taylor',
        'framework' => 'laravel'
    ],
    [
        'name' => 'andrew',
        'framework' => 'vue'
    ]
]);

$collection->forgetNested('name');

// [['framework' => 'laravel'],['framework' => 'vue']]
```

**Comment by ousid** [comment]
Hey, @aschmelyun 

You can achieve the same result with the following:

```php
$collection = collect([
    [
        'name' => 'taylor',
        'framework' => 'laravel'
    ],
    [
        'name' => 'andrew',
        'framework' => 'vue'
    ]
]);

$newCollection = [];

$collection->each(function ($element) use (&$newCollection) {
    unset($element['name']);
    $newCollection[] = $element;
});

// [['framework' => 'laravel'],['framework' => 'vue']]
```

But, I think it's a lot complicated than `$collection->forgetNested('name')`

Maybe @taylorotwell has another opinion?

**Reply by henzeb** [reply]
you could use $collection->map() instead. https://laravel.com/docs/9.x/collections#method-map

**Comment by Stevemoretz** [comment]
This is a great idea!
I was trying asterisks all over the place not knowing this wasn't even supported.
Obviously we can do it ourselves but it's just a great method to have.

**Reply by shaedrich** [reply]
> [Most methods](https://www.dankuck.com/2019/09/06/laravel-collection-dot-notation.html#methods-that-respect-dot-notation) that accept key names respect dot-notation. The list of [those that don’t](https://www.dankuck.com/2019/09/06/laravel-collection-dot-notation.html#methods-that-do-not-respect-dot-notation) is short: forget, get, has, only, prepend, put. Some other methods only accept callbacks, such as map.
>
> Any method that accepts the same parameters as where($key, $operator, $value) will also respect dot-notation.
>
> Only pluck and groupBy use the [array-wildcard *](https://www.dankuck.com/2019/09/06/laravel-collection-dot-notation.html#what-about-) correctly. The others will give wrong results.
&mdash; https://www.dankuck.com/2019/09/06/laravel-collection-dot-notation.html

Yeah, it's quite disappointing that Laravel's data, array and collection helpers are this inconsistent in that respect (especially that `data_get()` supports asterisks while `Arr::get()` does not).

**Comment by delhiarpitpatel** [comment]
This is a great idea!

**Comment by simonbuehler** [comment]
one thing that i'm missing in laravel is the power of using something like a [Hashpath Syntax](https://book.cakephp.org/4/en/core-libraries/hash.html#hash-path-syntax) from cakephp in collections, this is a real power feature not available in laravel and would allow to extract data from nested API responses like twitter into a handy collection.

**Comment by DudeMason** [comment]
This would be awesome. Also allowing it to remove all items with that key, as far as the nest goes! Not just the first or second level.

**Reply by DudeMason** [reply]
I found this thread because I had an instance today where it would have been nice to have this!

**Comment by rossjcooper** [comment]
I've added a PR to support this using dot notation and nested keys https://github.com/laravel/framework/pull/48614

**Comment by projct1** [comment]
Its actual for me still...
